### PR TITLE
Update clippy for latest rustc changes

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Specifically, this revises the clippy integration to utilize a new
callback to register its lints, as the prior editing of lint store in
Session is no longer possible.

---

changelog: none

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
This is for https://github.com/rust-lang/rust/pull/65193. cc @Manishearth

**Comment by Mark-Simulacrum** [comment]
This will need another pass since we need to `register_lints` as well for all the lints -- if you have thoughts on how to do that I'd be happy to do so. Previously this was implicit but that's no longer the case.

**Comment by Manishearth** [comment]
Got a try build we can rustup-toolchain-install-master to test with?

**Comment by Manishearth** [comment]
A lot of lib.rs is autogenerated with update-lints, so we can update that to do whatever registry you need

**Comment by Mark-Simulacrum** [comment]
Attempted to update update-lints to do the right thing and pushed relevant changes. Working on a try build in the rust PR.

File: clippy_dev/src/lib.rs [hunk 1]
```diff
@@ -136,6 +136,19 @@ pub fn gen_deprecated(lints: &[Lint]) -> Vec<String> {
         .collect::<Vec<String>>()
 }
 
+pub fn gen_register_lint_list(lints: &[Lint]) -> Vec<String> {
+    let pre = "    store.register_lints(&[".to_string();
+    let post = "    ]);".to_string();
+    let mut inner = lints
+        .iter()
+        .filter(|l| !(l.is_internal() || l.deprecation.is_some()))
+        .map(|l| format!("        &{}::{},", l.module, l.name.to_uppercase()))
```
Comment (flip1995) [inline clippy_dev/src/lib.rs]: I'd also sort this list.
```suggestion
        .map(|l| format!("        &{}::{},", l.module, l.name.to_uppercase()))
        .sorted()
```

**Comment by matthiaskrgr** [comment]
https://github.com/rust-lang/rust/pull/65193 was merged so we need this now

**Comment by lzutao** [comment]
You need to revert 974a05b806954ebdc6640b7ef539667bd7baca56 to rustup rustc-dev component.

**Comment by flip1995** [comment]
Oh so bors decided to finally merge this PR :smile: 

On it :+1:

**Comment by flip1995** [comment]
Now we should be good (at least `cargo test` past locally)

**Comment by Mark-Simulacrum** [comment]
(Let me know if you need anything from me here)

**Comment by lzutao** [comment]
CI passed.

**Pull request merged**